### PR TITLE
Create scalar UDFs as STABLE functions by default

### PR DESCRIPTION
**Pull request by skkyal**
Signed-off-by: Shlok Kumar Kyal ([skkyal@amazon.com](mailto:skkyal@amazon.com))

### Description
Earlier the tsql functions were created with volatility as `volatile` by default. From behaviour and performance perspective function should be created as 'stable' by default. Also, When a function is `stable`, it cannot have DDL or DML inside it. But it possible to DML only in table variables in tsql. 
With this PR made the default volatility during tsql function creation as `stable` except for the cases when a function contains a table variable (tvp or tvf).

The functions created before this commit will be by default created as `VOLATILE` and after this commit will be created as `STABLE` by default. Due to this as a part of this PR, we needed to make changes in JDBC and m/MVU (add version specific test)  test files related to tests`sp_babelfish_volatility` and `BABEL-2877` as in these tests volatility of functions are being checked.

### Issues Resolved
BABEL-3700

### Test Scenarios Covered ###
* **Use case based - Yes**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases - Yes**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y [hunk 1]
```diff
@@ -3616,12 +3616,13 @@ tsql_CreateFunctionStmt:
 						DefElem *lang = makeDefElem("language", (Node *) makeString("pltsql"), @1);
 						DefElem *body = makeDefElem("as", (Node *) list_make1(makeString($10)), @10);
 						DefElem *location = makeDefElem("location", (Node *) makeInteger(@4), @4);
+						DefElem *vol = makeDefElem("volatility", (Node *) makeString("stable"), @1);
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:3628]: Please add a comment here

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 2]
```diff
@@ -4503,6 +4514,45 @@ pltsql_validator(PG_FUNCTION_ARGS)
 		}
 
 		ReleaseSysCache(tuple);
+		
+		/* 
+		 * If the function has TVP in its arguments or function body 
+		 * it should be declared as VOLATILE by default 
+		 */
+		if(prokind == PROKIND_FUNCTION && (has_table_var && !is_itvf && !is_mstvf))
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pl_handler.c:4771]: Why arent we changing it for mstvf and istvf??

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 3]
```diff
@@ -4503,6 +4514,45 @@ pltsql_validator(PG_FUNCTION_ARGS)
...
+		{
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pl_handler.c:4772]: Why are we doing this in validator

**Comment by KushaalShroff** [comment]
Please add to the description which test files you had to change with a brief explanation

File: test/JDBC/expected/create_function_default_stable-vu-prepare.out [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+go
+
+create table #test_create_function_default_stable_tmp(a int)
+go
+
+create type test_create_function_default_stable_type AS TABLE(a int)
+go
+
+-- create functions not having tvp or tvf
+create function test_create_function_default_stable_f1()
```
Comment (KushaalShroff) [inline test/JDBC/expected/create_function_default_stable-vu-prepare.out]: What purpose is this test case signifying?

File: test/JDBC/expected/create_function_default_stable-vu-prepare.out [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+create function test_create_function_default_stable_tvp_f5(@tvp test_create_function_default_stable_type readonly) 
+returns int as begin insert into @tvp values(1)  return 0;end
+go
+
+-- create functions with tvf
+create function test_create_function_default_stable_tvf_f1() returns table as return (SELECT * from test_create_function_default_stable_t1)
+go
+
+create function test_create_function_default_stable_tvf_f2(@a int)
```
Comment (KushaalShroff) [inline test/JDBC/expected/create_function_default_stable-vu-prepare.out:52]: Please test nested functions as well with various pnc's.
eg: Normal function calling TVF

File: test/JDBC/expected/create_function_default_stable-vu-prepare.out [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+CREATE OR REPLACE FUNCTION test_create_function_default_stable_psql_f9() RETURNS int AS $$
+BEGIN
+	DECLARE @a int = 1;
+	declare @tvp table(a int);
+	RETURN @a
+END;
+$$ LANGUAGE pltsql STABLE;
+go
+
+CREATE OR REPLACE FUNCTION test_create_function_default_stable_psql_f10() RETURNS int AS $$
```
Comment (KushaalShroff) [inline test/JDBC/expected/create_function_default_stable-vu-prepare.out:261]: Add a negative test case for pltsql function in pg dialect returning record

**Comment by KushaalShroff** [comment]
Also please add to your description about the m/MVU impact.

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 2 above)
Comment (skkyal) [reply contrib/babelfishpg_tsql/src/pl_handler.c:4771]: in `gram-tsql-rule.y` there is separate rules for Inline Table Valued Function(itvf), Multi Statement Table Valued Function(mstvf) and Normal create Function.
So, in `gram-tsql-rule.y` I have added an option of `STABLE` only for normal create function rule. 
So for itvf and mstvf no volatility option is provided, hence it will be `VOLATILE` by default

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 3 above)
Comment (skkyal) [reply contrib/babelfishpg_tsql/src/pl_handler.c:4772]: So the variables related to table variables are set inside `do_compile` function. Specially inside `antlr_parser_cpp(proc_source)`. So actually the function is already created by that time. So I am updating the tuple of the function.

bt when a function is created 
```
do_compile
pltsql_compile
pltsql_validator
fmgr_security_definer 
FunctionCall1Coll
OidFunctionCall1Coll
ProcedureCreate
CreateFunction 
bbf_ProcessUtility
```

The main reason I use do changes inside `pltsql_validator` because we have access to relevant variable to identify table variable here, This function attempts to validate a PL/tsql function at CREATE FUNCTION time and no engine changes required.

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 7]
```diff
@@ -2142,7 +2142,7 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
 				{
 					if (isCompleteQuery)
 						EventTriggerDDLCommandStart(parsetree);
-
+
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: unnecessary change

**Review by KushaalShroff — DISMISSED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 3 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/pl_handler.c:4772]: Is it possible to do it in post_object_create hook? I am assuming it won't because it will be executed directly after CreateFunction/ProcedureCreate but it'll be better to verify.

**Review by HarshLunagariya — DISMISSED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 3 above)
Comment (skkyal) [reply contrib/babelfishpg_tsql/src/pl_handler.c:4772]: It won't be possible to do it in post_object_create as we cannot get relevant variable to get presence of tvp.

**Review by KushaalShroff — APPROVED** [review]
